### PR TITLE
chore: make parent scope explicit

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: packages/playwright-core/src/server/dispatchers/frameDispatcher.ts [hunk 1]
```diff
@@ -62,7 +62,7 @@ export class FrameDispatcher extends Dispatcher<Frame, channels.FrameChannel> im
         return;
       const params = { url: event.url, name: event.name, error: event.error ? event.error.message : undefined };
       if (event.newDocument)
-        (params as any).newDocument = { request: RequestDispatcher.fromNullable(this._scope, event.newDocument.request || null) };
+        (params as any).newDocument = { request: RequestDispatcher.fromNullable(scope.parentScope()!, event.newDocument.request || null) };
```
Comment (dgozman) [inline packages/playwright-core/src/server/dispatchers/frameDispatcher.ts:65]: This looks like a semantic change.

File: packages/playwright-core/src/server/dispatchers/jsHandleDispatcher.ts [hunk 2]
```diff
@@ -16,15 +16,19 @@
...
-import type { DispatcherScope } from './dispatcher';
 import { Dispatcher } from './dispatcher';
 import { ElementHandleDispatcher } from './elementHandlerDispatcher';
 import { parseSerializedValue, serializeValue } from '../../protocol/serializers';
+import type { PageDispatcher, WorkerDispatcher } from './pageDispatcher';
+import type { ElectronApplicationDispatcher } from './electronDispatcher';
+import type { BrowserContextDispatcher } from './browserContextDispatcher';
 
-export class JSHandleDispatcher extends Dispatcher<js.JSHandle, channels.JSHandleChannel> implements channels.JSHandleChannel {
+export type JSHandleDispatcherParentScope = BrowserContextDispatcher | PageDispatcher | WorkerDispatcher | ElectronApplicationDispatcher;
```
Comment (dgozman) [inline packages/playwright-core/src/server/dispatchers/jsHandleDispatcher.ts:26]: How `BrowserContextDispatcher` could be a parent of a handle?

File: packages/playwright-core/src/server/dispatchers/networkDispatchers.ts [hunk 3]
```diff
@@ -19,28 +19,30 @@ import type { APIRequestContext } from '../fetch';
...
+  static fromNullable(scope: BrowserContextDispatcher, request: Request | null): RequestDispatcher | undefined {
     return request ? RequestDispatcher.from(scope, request) : undefined;
   }
 
-  private constructor(scope: DispatcherScope, request: Request) {
+  private constructor(scope: BrowserContextDispatcher, request: Request) {
     const postData = request.postDataBuffer();
     super(scope, request, 'Request', {
-      frame: FrameDispatcher.fromNullable(scope, request.frame()),
+      frame: FrameDispatcher.fromNullable(scope as any as PageDispatcher, request.frame()), // this will get rerouted to the right parent.
```
Comment (dgozman) [inline packages/playwright-core/src/server/dispatchers/networkDispatchers.ts:45]: Oh my!

File: packages/playwright-core/src/server/dispatchers/pageDispatcher.ts [hunk 4]
```diff
@@ -154,7 +155,7 @@ export class PageDispatcher extends Dispatcher<Page, channels.PageChannel> imple
       return;
     }
     await this._page.setClientRequestInterceptor((route, request) => {
-      this._dispatchEvent('route', { route: RouteDispatcher.from(this._scope, route), request: RequestDispatcher.from(this._scope, request) });
+      this._dispatchEvent('route', { route: RouteDispatcher.from(this._scope, route, RequestDispatcher.from(this.parentScope()!, request)) });
```
Comment (dgozman) [inline packages/playwright-core/src/server/dispatchers/pageDispatcher.ts:158]: Again, semantic change?

**Review by dgozman — APPROVED** [review]

File: packages/playwright-core/src/server/dispatchers/browserContextDispatcher.ts [hunk 5]
```diff
@@ -34,15 +34,15 @@ import * as path from 'path';
...
+export class BrowserContextDispatcher extends Dispatcher<BrowserContext, channels.BrowserContextChannel, DispatcherScope, BrowserContextDispatcher> implements channels.BrowserContextChannel {
   _type_EventTarget = true;
   _type_BrowserContext = true;
   private _context: BrowserContext;
 
   constructor(parentScope: DispatcherScope, context: BrowserContext) {
     // We will reparent these to the context below.
-    const requestContext = APIRequestContextDispatcher.from(parentScope, context.fetchRequest);
-    const tracing = TracingDispatcher.from(parentScope, context.tracing);
+    const requestContext = APIRequestContextDispatcher.from(parentScope as BrowserContextDispatcher, context.fetchRequest);
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply packages/playwright-core/src/server/dispatchers/browserContextDispatcher.ts:44]: When we are performing the 'adopt' dance, we are always casting the parent scope to the eventual parent.

File: packages/playwright-core/src/server/dispatchers/frameDispatcher.ts (hunk 1 above)
Comment (pavelfeldman) [reply packages/playwright-core/src/server/dispatchers/frameDispatcher.ts:65]: Yes, it was a bug - we were using frame as a parent scope, whereas it should have been the context. Frame has no scope, so its parent is context.

File: packages/playwright-core/src/server/dispatchers/jsHandleDispatcher.ts (hunk 2 above)
Comment (pavelfeldman) [reply packages/playwright-core/src/server/dispatchers/jsHandleDispatcher.ts:26]: That would be a BindingCallDispatcher's handle. Let me try rerouting it to the page. It'll be a semantics change though.

File: packages/playwright-core/src/server/dispatchers/pageDispatcher.ts (hunk 4 above)
Comment (pavelfeldman) [reply packages/playwright-core/src/server/dispatchers/pageDispatcher.ts:158]: I don't think so.

File: packages/playwright-core/src/server/dispatchers/pageDispatcher.ts [hunk 6]
```diff
@@ -291,19 +296,20 @@ export class PageDispatcher extends Dispatcher<Page, channels.PageChannel> imple
...
     return result || new WorkerDispatcher(scope, worker);
   }
 
-  _type_Worker = true;
-  constructor(scope: DispatcherScope, worker: Worker) {
+  constructor(scope: PageDispatcher | BrowserContextDispatcher, worker: Worker) {
     super(scope, worker, 'Worker', {
       url: worker.url()
-    });
+    }, true);
```
Comment (dgozman) [inline packages/playwright-core/src/server/dispatchers/pageDispatcher.ts:312]: It seems like we can declare every object a scope now. Do we have an example where it's not a scope and it matters?

File: packages/playwright-core/src/server/dispatchers/frameDispatcher.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply packages/playwright-core/src/server/dispatchers/frameDispatcher.ts:65]: No, we are building the right scope based on the context scope inside the called method.

File: packages/playwright-core/src/server/dispatchers/networkDispatchers.ts [hunk 7]
```diff
@@ -62,22 +65,23 @@ export class RequestDispatcher extends Dispatcher<Request, channels.RequestChann
...
     return response ? ResponseDispatcher.from(scope, response) : undefined;
   }
 
-  private constructor(scope: DispatcherScope, response: Response) {
+  private constructor(contextScope: BrowserContextDispatcher, response: Response) {
+    const scope = parentScopeForRequest(contextScope, response.request());
     super(scope, response, 'Response', {
       // TODO: responses in popups can point to non-reported requests.
-      request: RequestDispatcher.from(scope, response.request()),
+      request: RequestDispatcher.from(contextScope, response.request()),
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply packages/playwright-core/src/server/dispatchers/networkDispatchers.ts:84]: No, check out the called method.

File: packages/playwright-core/src/server/dispatchers/networkDispatchers.ts [hunk 8]
```diff
@@ -19,36 +19,39 @@ import type { APIRequestContext } from '../fetch';
...
+      frame: request.frame() ? scope : undefined,
+      serviceWorker: request.serviceWorker() ? scope : undefined,
       url: request.url(),
       resourceType: request.resourceType(),
       method: request.method(),
       postData: postData === null ? undefined : postData,
       headers: request.headers(),
       isNavigationRequest: request.isNavigationRequest(),
-      redirectedFrom: RequestDispatcher.fromNullable(scope, request.redirectedFrom()),
+      redirectedFrom: RequestDispatcher.fromNullable(contextScope, request.redirectedFrom()),
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply packages/playwright-core/src/server/dispatchers/networkDispatchers.ts:54]: Not really...

File: packages/playwright-core/src/server/dispatchers/pageDispatcher.ts (hunk 6 above)
Comment (pavelfeldman) [reply packages/playwright-core/src/server/dispatchers/pageDispatcher.ts:312]: I might be able to do that in a follow-up.

**Pull request merged**